README.md
koa-static-ts
=================
[![npm version](https://badge.fury.io/js/koa-static-ts.svg)](https://badge.fury.io/js/koa-static-ts)
[![Build Status](https://travis-ci.org/HKUST-VISLab/koa-static-ts.svg?branch=master)](https://travis-ci.org/HKUST-VISLab/koa-static-ts)
[![codecov](https://codecov.io/gh/HKUST-VISLab/koa-static-ts/branch/master/graph/badge.svg)](https://codecov.io/gh/HKUST-VISLab/koa-static-ts)
[![David](https://david-dm.org/HKUST-VISLab/koa-static-ts/status.svg)](https://github.com/HKUST-VISLab/koa-static-ts)
[![Greenkeeper badge](https://badges.greenkeeper.io/HKUST-VISLab/koa-static-ts.svg)](https://greenkeeper.io/)

## Licences

[MIT](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.12](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.11...v0.1.12) (2019-05-25)


### Bug Fixes

* **package:** update @types/node to version 10.0.3 ([91cd71c](https://github.com/HKUST-VISLab/koa-static-ts/commit/91cd71c))
* **package:** update @types/node to version 11.9.0 ([5f1a021](https://github.com/HKUST-VISLab/koa-static-ts/commit/5f1a021))
* **package:** update @types/node to version 9.6.5 ([23b36ea](https://github.com/HKUST-VISLab/koa-static-ts/commit/23b36ea))


### Tests

* update some test ([5fa3fd2](https://github.com/HKUST-VISLab/koa-static-ts/commit/5fa3fd2))



<a name="0.1.11"></a>
## [0.1.11](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.10...v0.1.11) (2018-04-11)



<a name="0.1.10"></a>
## [0.1.10](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.9...v0.1.10) (2017-12-31)


### Bug Fixes

* conflict ([871070a](https://github.com/HKUST-VISLab/koa-static-ts/commit/871070a))
* fix conflict ([c3c4a3a](https://github.com/HKUST-VISLab/koa-static-ts/commit/c3c4a3a))



<a name="0.1.9"></a>
## [0.1.9](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.8...v0.1.9) (2017-12-31)



<a name="0.1.8"></a>
## [0.1.8](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.7...v0.1.8) (2017-07-30)


### Bug Fixes

* default export ([86f39f6](https://github.com/HKUST-VISLab/koa-static-ts/commit/86f39f6))



<a name="0.1.7"></a>
## [0.1.7](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.6...v0.1.7) (2017-07-30)


### Bug Fixes

* default export and Options Interface ([8fe8556](https://github.com/HKUST-VISLab/koa-static-ts/commit/8fe8556))



<a name="0.1.6"></a>
## [0.1.6](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.5...v0.1.6) (2017-07-30)



<a name="0.1.5"></a>
## [0.1.5](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.4...v0.1.5) (2017-07-30)



<a name="0.1.4"></a>
## [0.1.4](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.3...v0.1.4) (2017-07-29)


### Bug Fixes

* skip clean up after test ([b689d42](https://github.com/HKUST-VISLab/koa-static-ts/commit/b689d42))



<a name="0.1.3"></a>
## [0.1.3](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.2...v0.1.3) (2017-07-29)


### Bug Fixes

* use ctx.throw instead ([d187090](https://github.com/HKUST-VISLab/koa-static-ts/commit/d187090))



<a name="0.1.2"></a>
## [0.1.2](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.1...v0.1.2) (2017-07-27)



<a name="0.1.1"></a>
## [0.1.1](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.1.0...v0.1.1) (2017-07-27)


### Bug Fixes

* add type for args of  setHeader ([32e901b](https://github.com/HKUST-VISLab/koa-static-ts/commit/32e901b))
* change ceil maxAge to floor maxAge ([da2cf3f](https://github.com/HKUST-VISLab/koa-static-ts/commit/da2cf3f))



<a name="0.1.0"></a>
# [0.1.0](https://github.com/HKUST-VISLab/koa-static-ts/compare/v0.0.1...v0.1.0) (2017-07-21)


### Features

* unittest ([972fa1f](https://github.com/HKUST-VISLab/koa-static-ts/commit/972fa1f))



<a name="0.0.1"></a>
## 0.0.1 (2017-07-21)


### Bug Fixes

* add `strict` to tsconfig ([bb80b31](https://github.com/HKUST-VISLab/koa-static-ts/commit/bb80b31))
* add utils ([605145f](https://github.com/HKUST-VISLab/koa-static-ts/commit/605145f))


### Features

* init, convert to ts ([3683ffc](https://github.com/HKUST-VISLab/koa-static-ts/commit/3683ffc))
* init. ([ae9ef75](https://github.com/HKUST-VISLab/koa-static-ts/commit/ae9ef75))


test/send.spec.ts
import test from 'ava';
import * as Koa from 'koa';
import * as path from 'path';
import * as request from 'supertest';
import { send } from '../src/send';

const decompressSync = require('iltorb').decompressSync;

interface TestContext {
    app: Koa;
}

test.beforeEach('init app', t => {
    const app = new Koa();
    app.use(async (ctx, next) => {
        try {
            await next();
        } catch (err) {
            // will only respond with JSON
            ctx.status = err.statusCode || err.status || 500;
            ctx.response.status = ctx.status;
            ctx.body = {
                message: err.message,
            };
        }
    });
    t.context = {
        app,
    };
});

test('send with no .root', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, '');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    let res = await req.get(`/${path.join(__dirname, '/fixtures/hello.txt')}`);
    t.is(res.status, 400, 'when the path is absolute should 400');

    res = await req.get('/test/fixtures/hello.txt');
    t.is(res.status, 200, 'when the path is relative should 200');
    t.deepEqual(res.text, 'world', 'when the path is relative should get the content');

    res = await req.get('/../fixtures/hello.txt');
    t.is(res.status, 403, 'when the path contains .. should 403');
});

test('send with .root', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test/fixtures');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    let res = await req.get(`/${path.join(__dirname, '/fixtures/hello.txt')}`);
    t.is(res.status, 400, 'when the path is absolute should 400');

    res = await req.get('/hello.txt');
    t.is(res.status, 200, 'when the path is relative and file exists should 200');
    t.is(res.text, 'world', 'when the path is relative should get the content');

    res = await req.get('/hello1.txt');
    t.is(res.status, 404, 'when the path is relative and does not exist should 404');

    res = await req.get('/../../package.json');
    t.is(res.status, 403, 'when the path resolves above the root should 403');

    res = await req.get('/../../test/fixtures/world/index.html');
    t.is(res.status, 403, 'when the path resolves within root but use .. should 403');
});

test('with .index', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/world/');
    t.is(res.status, 200, 'when the index file is present should 200');
    t.deepEqual(res.text, 'html index', 'when the index file is present should serve it, default to index.html');

});

test('with .index path', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test/fixtures/world');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/');
    t.is(res.status, 200, 'when the index file is present should 200');
    t.deepEqual(res.text, 'html index', 'when the index file is present should serve it, default to index.html');
});

test('with .index index.txt', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test/fixtures', { index: 'index.txt' });
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/');
    t.is(res.status, 200, 'when the index file is present should 200');
    t.deepEqual(res.text, 'text index', 'when the index file is present should serve it');
});

test('with .index false', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test/fixtures', { index: false });
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/');
    t.is(res.status, 404, 'when the index file option is set to false');
});

test('when path is not a file', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        if (ctx.path === '/test') {
            await send(ctx, '');
        } else if (ctx.path === '/') {
            await send(ctx, 'test', { format: false });
        }
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    let res = await req.get('/test');
    t.is(res.status, 404, 'when path is not a file should 404');

    res = await req.get('/');
    t.is(res.status, 404, 'should return undefined if format is set to false');
});

// test('when path is not a file', async (t) => {
//     const { app } = t.context as TestContext;
//     app.use(async (ctx) => {
//         await send(ctx, 'test', { format: false });
//     });
//     const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
//     const res = await req.get('/');
//     t.is(res.status, 404, 'should return undefined if format is set to false');
// });

test('when path is a directory', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test', { format: false });
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures');
    t.is(res.status, 404, 'when path is a directory should 404');
});

test('when path does not finish with slash and format is disabled', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test', { format: false });
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/world');
    t.is(res.status, 404, 'should 404');
});

test('when path does not finish with slash and format is enabled', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test', { format: true });
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/world')
        .expect('content-type', 'text/html; charset=utf-8')
        .expect('content-length', '10');
    t.is(res.status, 200, 'should 200');
});

test('when path does not finish with slash and index is disabled', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test', { index: false });
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/world');
    t.is(res.status, 404, 'should 404');
});

test('when path is malformed', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, '');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/%');
    t.is(res.status, 400, 'should 400');
});

test('when path is a file or .gz version when requested and if possible', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        const returnPath = await send(ctx, 'test');
        if (ctx.path.includes('gzip.json') && ctx.header['accept-encoding'].includes('gzip')) {
            t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path + '.gz'), 'should return the path');
        } else {
            t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path), 'should return the path');
        }
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    let res = await req.get('/fixtures/user.json');
    t.is(res.status, 200, 'should 200');

    res = await req.get('/fixtures/gzip.json')
        .set('Accept-Encoding', 'deflate, identity');
    t.deepEqual(res.header['content-length'], '18',
        'should return the original version when accept-encoding has no gzip or br');
    t.deepEqual(res.text, '{ "name": "tobi" }',
        'should return the original version when accept-encoding has no gzip or br');
    t.is(res.status, 200, 'should 200');

    res = await req.get('/fixtures/gzip.json')
        .set('Accept-Encoding', 'gzip, deflate, identity');
    t.deepEqual(res.header['content-length'], '48', 'should return .gz path (gzip option defaults to true)');
    t.deepEqual(res.text, '{ "name": "tobi" }', 'should return .gz path (gzip option defaults to true)');
    t.is(res.status, 200, 'should return .gz path (gzip option defaults to true)');
});

// test('when path is a file or .gz version when requested and if possible', async (t) => {
//     const { app } = t.context as TestContext;
//     app.use(async (ctx) => {
//         const returnPath = await send(ctx, 'test');
//         t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path + '.gz'), 'should return the path');
//     });
//     const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
//     const res = await req.get('/fixtures/gzip.json')
//         .set('Accept-Encoding', 'gzip, deflate, identity');
//     t.deepEqual(res.header['content-length'], '48', 'should return .gz path (gzip option defaults to true)');
//     t.deepEqual(res.text, '{ "name": "tobi" }', 'should return .gz path (gzip option defaults to true)');
//     t.is(res.status, 200, 'should return .gz path (gzip option defaults to true)');
// });

test('when path is a file or .gz version when requested and if possible with gzip false', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        const returnPath = await send(ctx, 'test', { gzip: false });
        t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path), 'should return the path');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/gzip.json')
        .set('Accept-Encoding', 'gzip, deflate, identity');
    t.deepEqual(res.header['content-length'], '18', 'should not return .gz path when gzip option is false');
    t.deepEqual(res.text, '{ "name": "tobi" }', 'should not return .gz path when gzip option is false');
    t.is(res.status, 200, 'should not return .gz path when gzip option is false');
});

test('when path is a file or .br version when requested and if possible', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        if (ctx.header['accept-encoding'].includes('gzip')) {
            const returnPath = await send(ctx, 'test');
            t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path + '.br'), 'should return the path');
        } else {
            const returnPath = await send(ctx, 'test', { brotli: false });
            t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path), 'should return the path');
        }
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    let res = await req.get('/fixtures/gzip.json')
        .set('Accept-Encoding', 'br, gzip, deflate, identity');
    t.deepEqual(res.header['content-length'], '22', 'should return .br path (brotli option defaults to true)');
    t.is(res.status, 200, 'should return .br path (brotli option defaults to true)');
    t.deepEqual(decompressSync(res.body).toString(), '{ "name": "tobi" }',
        'should return .br path (brotli option defaults to true)');

    res = await req.get('/fixtures/gzip.json')
        .set('Accept-Encoding', 'br, deflate, identity');
    t.deepEqual(res.header['content-length'], '18', 'should not return .br path when brotli option is false');
    t.deepEqual(res.text, '{ "name": "tobi" }', 'should not return .br path when brotli option is false');
    t.is(res.status, 200, 'should not return .br path when brotli option is false');
});

// test('when path is a file or .br version when requested and if possible', async (t) => {
//     const { app } = t.context as TestContext;
//     app.use(async (ctx) => {
//         const returnPath = await send(ctx, 'test', { brotli: false });
//         t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path), 'should return the path');
//     });
//     const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
//     const res = await req.get('/fixtures/gzip.json')
//         .set('Accept-Encoding', 'br, deflate, identity');
//     t.deepEqual(res.header['content-length'], '18', 'should not return .br path when brotli option is false');
//     t.deepEqual(res.text, '{ "name": "tobi" }', 'should not return .br path when brotli option is false');
//     t.is(res.status, 200, 'should not return .br path when brotli option is false');
// });

test('when path is a file or .br version when requested and if possible with brotli false', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        const returnPath = await send(ctx, 'test', { brotli: false });
        t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path + '.gz'), 'should return the path');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/gzip.json')
        .set('Accept-Encoding', 'br, gzip, deflate, identity');
    t.deepEqual(res.header['content-length'], '48', 'should return .gz path when brotli option is turned off');
    t.deepEqual(res.text, '{ "name": "tobi" }', 'should return .gz path when brotli option is turned off');
    t.is(res.status, 200, 'should return .gz path when brotli option is turned off');
});

test('when path is a file and max-age is set', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        const returnPath = await send(ctx, 'test', { maxAge: 5000 });
        t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path), 'should return the path');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/user.json');
    t.is(res.status, 200, 'should set max-age in seconds');
    t.deepEqual(res.header['cache-control'], 'max-age=5', 'should set max-age in seconds');
});

test('when path is a file  and max-age is set', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        const returnPath = await send(ctx, 'test', { maxAge: 1234 });
        t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path), 'should return the path');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/user.json');
    t.is(res.status, 200, 'should truncate fractional values for max-age');
    t.deepEqual(res.header['cache-control'], 'max-age=1', 'should truncate fractional values for max-age');
});

test('when path is a file and immutable is specified', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        const returnPath = await send(ctx, 'test', { maxAge: 31536000000, immutable: true });
        t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path), 'should return the path');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/user.json');
    t.is(res.status, 200, 'should set the immutable directive');
    t.deepEqual(res.header['cache-control'], 'max-age=31536000,immutable',
        'should set the immutable directive');
});

test('.immutable option when trying to get a non-existent file', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        const returnPath = await send(ctx, 'test', { immutable: true });
        t.deepEqual(returnPath, path.join(process.cwd(), 'test', ctx.path), 'should return the path');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/does-not-exist.json');
    t.is(res.status, 404, 'should set the immutable directive');
    t.is(res.header['cache-control'], undefined,
        'should not set the Cache-Control header');
});

test('.hidden option when trying to get a hidden file', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/.hidden');
    t.is(res.status, 404, 'should 404');
});

test('.hidden option when trying to get a file from a hidden directory', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/.private/id_rsa.txt');
    t.is(res.status, 404, 'should 404');
});

test('.hidden option when trying to get a hidden file and .hidden check is turned off', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test', { hidden: true });
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/.hidden');
    t.is(res.status, 200, 'should 200');
});

test('.extensions option when trying to get a file without extension with no .extensions sufficed', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/hello');
    t.is(res.status, 404, 'should 404');
});

test('.extensions option when trying to get a file without extension with no matching .extensions', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test', { extensions: ['json', 'htm', 'html'] });
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/hello');
    t.is(res.status, 404, 'should 404');
});

test('.extensions option when trying to get a file without extension with\
matching .extensions sufficed first matched should be sent', async (t) => {
        const { app } = t.context as TestContext;
        app.use(async (ctx) => {
            await send(ctx, 'test', { extensions: ['json', 'txt', 'html'] });
        });
        const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
        const res = await req.get('/fixtures/user');
        t.is(res.status, 200, 'should 200');
        t.deepEqual(res.header['content-type'], 'application/json; charset=utf-8', 'should 200 and application/json');
    });

test('.extensions option when trying to get a file without extension with matching .extensions sufficed', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test', { extensions: ['txt'] });
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/hello');
    t.is(res.status, 200, 'should 200');
});

test('.extensions option when trying to get a file without extension with\
matching doted .extensions sufficed', async (t) => {
        const { app } = t.context as TestContext;
        app.use(async (ctx) => {
            await send(ctx, 'test', { extensions: ['.txt'] });
        });
        const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
        const res = await req.get('/fixtures/hello');
        t.is(res.status, 200, 'should 200');
    });

test('it should set the content-type', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/user.json');
    t.is(res.status, 200, 'should 200');
    t.deepEqual(res.header['content-type'], 'application/json; charset=utf-8', 'should 200 and set content-type');
});

test('it should set the content-length', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/user.json');
    t.is(res.status, 200, 'should 200');
    t.deepEqual(res.header['content-length'], '18', 'should 200 and set content-length');
});

test('it should set Last-Modified', async (t) => {
    const { app } = t.context as TestContext;
    app.use(async (ctx) => {
        await send(ctx, 'test');
    });
    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/user.json');
    t.is(res.status, 200, 'should 200');
    t.truthy(res.header['last-modified'].indexOf('GMT'), 'should set last-modified');
});

test('with setHeaders should not edit already set headers', async (t) => {
    const { app } = t.context as TestContext;
    const testFilePath = '/fixtures/user.json';
    const normalizedTestFilePath = path.normalize(testFilePath);

    app.use(async (ctx) => {
        await send(ctx, 'test', {
            setHeaders(res_, path_, stats) {
                t.deepEqual(path_.substr(-normalizedTestFilePath.length), normalizedTestFilePath, 'path in setHeaders');
                t.is(stats.size, 18, 'stats in setHeaders');
                t.truthy(res_, 'res in setHeaders');

                // these can be set
                res_.setHeader('Cache-Control', 'max-age=0,must-revalidate');
                res_.setHeader('Last-Modified', 'foo');
                // this one can not
                res_.setHeader('Content-Length', '9000');
            },
        });
    });

    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    const res = await req.get('/fixtures/user.json');
    t.is(res.status, 200, 'should 200');
    t.deepEqual(res.header['cache-control'], 'max-age=0,must-revalidate',
        'should set cache-control');
    t.deepEqual(res.header['last-modified'], 'foo', 'should set last-modified');
    t.deepEqual(res.header['content-length'], '18', 'should set content-length');
});

test('should cleanup on socket error', async (t) => {
    const { app } = t.context as TestContext;
    let stream: any;
    app.use(async (ctx) => {
        await send(ctx, 'test');
        stream = ctx.body;
        ctx.socket.emit('error', new Error('boom'));
    });

    const req = request(app.listen(10000 + Math.ceil(Math.random() * 20000)));
    await t.throwsAsync(req.get('/fixtures/user.json'), 'socket hang up', 'should throw error');
    t.truthy(stream.destroyed, 'should destroyed the stream');
    // TODO, should return 500
    // t.is(res.status, 500, 'should 500');
});


src/index.ts
import { Context } from 'koa';
import { send, StaticServerOptions } from './send';

export default function serve(root: string = '.', opts?: StaticServerOptions) {

    if (opts && opts.defer) {
        return async (ctx: Context, next: () => Promise<any>) => {
            await next();

            if (ctx.method !== 'HEAD' && ctx.method !== 'GET') {
                return;
            }

            // response is already handled
            if ((ctx.body !== null && ctx.body !== undefined) || ctx.status !== 404) {
                return;
            }
            try {
                await send(ctx, root, opts);
            } catch (err) {
                if (err.status !== 404) {
                    throw err;
                }
            }
        };
    }

    return async (ctx: Context, next: () => Promise<any>) => {
        let done: string | false = false;

        if (ctx.method === 'HEAD' || ctx.method === 'GET') {
            try {
                done = await send(ctx, root, opts);
            } catch (err) {
                if (err.status !== 404) {
                    throw err;
                }
            }
        }

        if (!done) {
            await next();
        }
    };

}

export { StaticServerOptions };


src/utils.ts
import { exists as rawExists, stat as rawStat, Stats } from 'fs';
import { Context } from 'koa';
import { isAbsolute, normalize, resolve as pathResolve, sep } from 'path';

export async function exists(path: string | Buffer) {
    return new Promise<boolean>((resolve, reject) => {
        rawExists(path, (result: boolean) => {
            resolve(result);
        });
    });
}

export async function stat(path: string | Buffer) {
    return new Promise<Stats>((resolve, reject) => {
        rawStat(path, (err, stats) => {
            return err ? reject(err) : resolve(stats);
        });
    });
}

const UP_PATH_REGEXP = /(?:^|[\\/])\.\.(?:[\\/]|$)/;

/**
 * Resolve relative path against a root path
 *
 * @param {string} rootPath
 * @param {string} relativePath
 * @return {string}
 * @public
 */
export function resolvePath(ctx: Context, rootPath: string, relativePath: string) {
    let root = rootPath;

    // containing NULL bytes is malicious
    if (relativePath.indexOf('\0') !== -1) {
        ctx.throw(400, 'Malicious Path');
    }

    // path should never be absolute
    if (isAbsolute(relativePath) || isAbsolute(relativePath)) {
        ctx.throw(400, 'Malicious Path');
    }

    // path outside root
    if (UP_PATH_REGEXP.test(normalize('.' + sep + relativePath))) {
        ctx.throw(403);
    }

    // resolve & normalize the root path
    root = normalize(pathResolve(root) + sep);

    // resolve the path
    return pathResolve(root, relativePath);
}


src/send.ts
import { createReadStream, Stats } from 'fs';
import { ServerResponse } from 'http';
import { Context } from 'koa';
import { basename, extname, normalize, parse, resolve, sep } from 'path';
import { exists, resolvePath, stat } from './utils';

export interface StaticServerOptions {
    defer?: boolean;
    index?: string | false;
    maxAge?: number;
    immutable?: boolean;
    hidden?: boolean;
    format?: boolean;
    extensions?: string[] | false;
    brotli?: boolean;
    gzip?: boolean;
    setHeaders?: (res: ServerResponse, path: string, stat: Stats) => void;
}

const staticServerDefaultOptions = {
    index: 'index.html' as 'index.html',
    defer: false as false,
    maxAge: 0 as 0,
    immutable: false as false,
    hidden: false as false,
    format: false as false,
    extensions: false as false,
    brotli: true as false,
    gzip: true as true,
};

/**
 * Send file at `path` with the
 * given `options` to the koa `ctx`.
 *
 * @param {Context} ctx
 * @param {String} path
 * @param {Object} [opts]
 * @return {Function}
 * @api public
 */
export async function send(ctx: Context, rootPath: string,
                           opts: StaticServerOptions = {}) {

    const root = normalize(resolve(rootPath));
    const { index,
        maxAge,
        immutable,
        hidden,
        format,
        extensions,
        brotli,
        gzip,
        setHeaders,
    } = Object.assign({}, staticServerDefaultOptions, opts);

    let path: string = ctx.path.substr(parse(ctx.path).root.length);

    // normalize path
    try {
        path = decodeURIComponent(path);
    } catch  (err) {
        ctx.throw(400, 'faild to decoded', err);
    }

    // index file support
    const trailingSlash = ctx.path[ctx.path.length - 1] === '/';
    if (index && trailingSlash) {
        path += index;
    }
    path = resolvePath(ctx, root, path);

    // hidden file support, ignore
    if (!hidden && isHidden(root, path)) {
        return false;
    }

    // serve brotli file when possible otherwise gzipped file when possible
    if (ctx.acceptsEncodings('br', 'deflate', 'identity') === 'br' &&
        brotli &&
        (await exists(path + '.br'))) {
        path = path + '.br';
        ctx.set('Content-Encoding', 'br');
        ctx.res.removeHeader('Content-Length');
    } else if (ctx.acceptsEncodings('gzip', 'deflate', 'identity') === 'gzip' &&
        gzip &&
        (await exists(path + '.gz'))) {

        path = path + '.gz';
        ctx.set('Content-Encoding', 'gzip');
        ctx.res.removeHeader('Content-Length');
    }

    if (extensions && !/\..*$/.exec(path)) {
        for (let ext of extensions) {
            if (!/^\./.exec(ext)) {
                ext = '.' + ext;
            }
            if (await exists(path + ext)) {
                path = path + ext;
                break;
            }
        }
    }

    // stat
    let stats: Stats;
    try {
        stats = await stat(path);

        // Format the path to serve static file servers
        // and not require a trailing slash for directories,
        // so that you can do both `/directory` and `/directory/`
        if (stats.isDirectory()) {
            if (format && index) {
                path += '/' + index;
                stats = await stat(path);
            } else {
                return false;
            }
        }
    } catch (err) {
        const notfound = new Set(['ENOENT', 'ENAMETOOLONG', 'ENOTDIR']);
        if (notfound.has(err.code)) {
            ctx.throw(404, err);
        }
        ctx.throw(500, err);
        return false; // TODO remove when typescript fix the control flow detect of never return type
    }

    if (setHeaders) {
        setHeaders(ctx.res, path, stats);
    }

    // stream
    ctx.set('Content-Length', stats.size.toString());
    if (!ctx.response.get('Last-Modified')) {
        ctx.set('Last-Modified', stats.mtime.toUTCString());
    }
    if (!ctx.response.get('Cache-Control')) {
        const directives = ['max-age=' + Math.floor(maxAge / 1000)];
        if (immutable) {
            directives.push('immutable');
        }
        ctx.set('Cache-Control', directives.join(','));
    }
    ctx.type = type(path);
    ctx.body = createReadStream(path);

    return path;
}

/**
 * Check if it's hidden.
 *
 * @param {string} root
 * @param {string} path
 * @returns
 */
function isHidden(root: string, path: string) {
    const filePath = path.substr(root.length).split(sep);
    for (let i = 0, len = filePath.length; i < len; ++i) {
        if (filePath[i][0] === '.') {
            return true;
        }
    }
    return false;
}

/**
 * File type.
 *
 * @param {string} file
 * @returns
 */
function type(file: string) {
    return extname(basename(file, '.gz'));
}
